### PR TITLE
Glimpse: Convert generateBp into a composite build

### PR DESCRIPTION
**Pull request by SebaUbuntu**
* We share this code within our apps, extract it into a common package and decommonize whatever might be app-specific

Change-Id: I403a0d8ce4c78b9a6ed2001e20611926c608ae06

### CODE REVIEW

File: app/build.gradle.kts [hunk 1]
```diff
@@ -90,279 +94,21 @@ dependencies {
...
-
-                        """.trimIndent().format(formatDeps(true))
-                    )
-                }
-
-                else -> throw Exception("Unknown file extension: ${it.extension}")
-            }
-        }
-    }
+    ).run()
```
Comment (luca020400) [inline app/build.gradle.kts]: can now use () instead

File: generateBpPlugin/src/main/java/org/lineageos/generatebp/GenerateBp.kt [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+    private val project: Project,
+    private val android: BaseAppModuleExtension,
+    private val isAvailableInAOSP: (group: String, artifactId: String) -> Boolean,
+    private val defaultMinSdkVersion: Int = DEFAULT_MIN_SDK_VERSION,
+) {
+    private val configuration = project.configurations["releaseRuntimeClasspath"]
+
+    private val libsBase = File("${project.projectDir.absolutePath}/libs")
+
+    fun run() {
```
Comment (luca020400) [inline generateBpPlugin/src/main/java/org/lineageos/generatebp/GenerateBp.kt]: operator fun invoke() {

File: app/build.gradle.kts (hunk 1 above)
Comment (luk1337) [reply app/build.gradle.kts]: done.

File: generateBpPlugin/src/main/java/org/lineageos/generatebp/GenerateBp.kt (hunk 2 above)
Comment (luk1337) [reply generateBpPlugin/src/main/java/org/lineageos/generatebp/GenerateBp.kt]: done.

**Comment by luk1337** [comment]
Should we move generateBpPlugin to app/generateBpPlugin since that's what uses it anyway?

**Comment by SebaUbuntu** [comment]
> Should we move generateBpPlugin to app/generateBpPlugin since that's what uses it anyway?

If we can do that sure

**Comment by luk1337** [comment]
> > Should we move generateBpPlugin to app/generateBpPlugin since that's what uses it anyway?
> 
> If we can do that sure

Will try then

**Comment by luca020400** [comment]
Since it'll be common maybe we can have a .gitsubmodule?
( I know AOSP doesn't like it, but we can import the code? )

**Comment by SebaUbuntu** [comment]
> > > Should we move generateBpPlugin to app/generateBpPlugin since that's what uses it anyway?
> > 
> > 
> > If we can do that sure
> 
> Will try then

Now that I think about it, I guess it could live in project root, we might need generateBp outside main app (e.g. something like lens_launcher), especially with that `includeBuild("app/generateBpPlugin")` in root settings.gradle.kts...

> Since it'll be common maybe we can have a .gitsubmodule? ( I know AOSP doesn't like it, but we can import the code? )

^, would be nice if repo supports submodules

**Comment by luk1337** [comment]
> > > > Should we move generateBpPlugin to app/generateBpPlugin since that's what uses it anyway?
> > > 
> > > 
> > > If we can do that sure
> > 
> > 
> > Will try then
> 
> Now that I think about it, I guess it could live in project root, we might need generateBp outside main app (e.g. something like lens_launcher), especially with that `includeBuild("app/generateBpPlugin")` in root settings.gradle.kts...
> 
> > Since it'll be common maybe we can have a .gitsubmodule? ( I know AOSP doesn't like it, but we can import the code? )
> 
> ^, would be nice if repo supports submodules

Wouldn't we prefer to publish the plugin on maven repo?

**Comment by luca020400** [comment]
> > > > Should we move generateBpPlugin to app/generateBpPlugin since that's what uses it anyway?
> > > 
> > > 
> > > If we can do that sure
> > 
> > Will try then
> 
> Now that I think about it, I guess it could live in project root, we might need generateBp outside main app (e.g. something like lens_launcher), especially with that `includeBuild("app/generateBpPlugin")` in root settings.gradle.kts...
> 
> > Since it'll be common maybe we can have a .gitsubmodule? ( I know AOSP doesn't like it, but we can import the code? )
> 
> ^, would be nice if repo supports submodules

It does apparently? Since when
Adding sync-s="true" to the entry will do that?!??

**Comment by luk1337** [comment]
rip

**Pull request closed without merging**